testsupport/signup_request.go

```diff
@@ -0,0 +1,270 @@
+	}
+}
+
+type signupRequest struct {
+	t                    *testing.T
+	hostAwait            *wait.HostAwaitility
+	memberAwait          *wait.MemberAwaitility
+	member2Await         *wait.MemberAwaitility
+	ensureMUR            bool
+	manuallyApprove      bool
+	verificationRequired bool
+	username             *string
+	email                *string
+	requiredHTTPStatus   int
+	targetCluster        *wait.MemberAwaitility
+	conditions           []toolchainv1alpha1.Condition
+	userSignup           *toolchainv1alpha1.UserSignup
+	mur                  *toolchainv1alpha1.MasterUserRecord
+}
+
+func (r *signupRequest) Username(username string) SignupRequest {
+	r.username = &username
+	return r
+}
+
+func (r *signupRequest) Email(email string) SignupRequest {
+	r.email = &email
+	return r
+}
+
+func (r *signupRequest) Resources() (*toolchainv1alpha1.UserSignup, *toolchainv1alpha1.MasterUserRecord) {
+	return r.userSignup, r.mur
+}
+
+func (r *signupRequest) EnsureMUR() SignupRequest {
+	r.ensureMUR = true
+	return r
+}
+
+func (r *signupRequest) ManuallyApprove() SignupRequest {
+	r.manuallyApprove = true
+	return r
+}
+
+func (r *signupRequest) RequireConditions(conditions ...toolchainv1alpha1.Condition) SignupRequest {
+	r.conditions = conditions
+	return r
+}
+
+func (r *signupRequest) VerificationRequired() SignupRequest {
+	r.verificationRequired = true
+	return r
+}
+
+func (r *signupRequest) TargetCluster(targetCluster *wait.MemberAwaitility) SignupRequest {
+	r.targetCluster = targetCluster
+	return r
+}
+
+func (r *signupRequest) RequireHTTPStatus(httpStatus int) SignupRequest {
+	r.requiredHTTPStatus = httpStatus
+	return r
+}
+
+func (r *signupRequest) Execute() SignupRequest {
+	WaitUntilBaseNSTemplateTierIsUpdated(r.t, r.hostAwait)
+
+	var username string
+	if r.username != nil {
+		username = *r.username
+	} else {
+		username = fmt.Sprintf("testuser-%s", uuid.Must(uuid.NewV4()).String())
+	}
+
+	// Create a token and identity to sign up with
+	userIdentity := &authsupport.Identity{
+		ID:       uuid.Must(uuid.NewV4()),
+		Username: username,
+	}
+
+	var email string
+	if r.email != nil {
+		email = *r.email
+	} else {
+		email = fmt.Sprintf("%s@test.com", username)
+	}
+
+	emailClaim0 := authsupport.WithEmailClaim(email)
+	token0, err := authsupport.GenerateSignedE2ETestToken(*userIdentity, emailClaim0)
+	require.NoError(r.t, err)
+
+	// Call the signup endpoint
+	invokeEndpoint(r.t, "POST", r.hostAwait.RegistrationServiceURL+"/api/v1/signup",
+		token0, "", r.requiredHTTPStatus)
+
+	// Wait for the UserSignup to be created
+	userSignup, err := r.hostAwait.WaitForUserSignup(userIdentity.ID.String())
+	require.NoError(r.t, err)
+
+	if r.targetCluster != nil {
+		if r.hostAwait.GetToolchainConfig().Spec.Host.AutomaticApproval.Enabled != nil {
+			require.False(r.t, *r.hostAwait.GetToolchainConfig().Spec.Host.AutomaticApproval.Enabled,
+				"cannot specify a target cluster for new signup requests while automatic approval is enabled")
```

Comment:
The test might still pass if during user provisioning the correct cluster happens to be randomly selected, however if not, and the test is actually checking for which cluster the user was provisioned in (like some tests do, which are the reason I added this error message) then it would fail, and with a much more cryptic error.

Reply:
ok, fair enough :+1: